client/network/src/request_responses.rs

```diff
@@ -497,21 +586,18 @@ pub enum RegisterError {
 	DuplicateProtocol(#[error(ignore)] Cow<'static, str>),
 }
 
-/// Error when sending a request.
+/// Error in a request.
 #[derive(Debug, derive_more::Display, derive_more::Error)]
-pub enum SendRequestError {
+pub enum RequestFailure {
 	/// We are not currently connected to the requested peer.
 	NotConnected,
 	/// Given protocol hasn't been registered.
 	UnknownProtocol,
-}
-
-/// Error in a request.
-#[derive(Debug, derive_more::Display, derive_more::Error)]
-pub enum RequestFailure {
 	/// Remote has closed the substream before answering, thereby signaling that it considers the
 	/// request as valid, but refused to answer it.
 	Refused,
+	/// The remote replied, but the local node is no longer interested in the response.
+	Obsolete,
```

Comment:
```suggestion
	/// When the response arrived the request was already aborted by the local node.
	Aborted,
```
It doesn't really sound like a failure condition though. Does it need to be reported?

Reply:
Whether it is a failure condition or not, I am not sure. My first intuition was yes.

I do think it should be reported on the statistics code path. E.g. say we have a lot of outgoing requests per second the fact that a high percentage of those become aborted during their lifetime would directly point to where to optimize.

Obviously the component triggering the initial request won't receive it, as it is no longer interested in the request in the first place.

Reply:
> I do think it should be reported on the statistics code path.

Yes, I guess that also comes down to the fact that we don't want in-place metrics collection throughout the code base. cf. https://github.com/paritytech/substrate/pull/7478#discussion_r528781937